_stbt/core.py

```diff
@@ -647,29 +671,27 @@ def match(self, image, frame=None, match_parameters=None,
         with numpy_from_sample(frame, readonly=True) as npframe:
             region = Region.intersect(_image_region(npframe), region)
 
-            matched, match_region, first_pass_certainty = _match(
-                _crop(npframe, region), template.image, match_parameters,
-                imglog)
-
-            match_region = Region.from_extents(*match_region).translate(
-                region.x, region.y)
-            result = MatchResult(
-                get_frame_timestamp(frame), matched, match_region,
-                first_pass_certainty, numpy.copy(npframe),
-                (template.name or template.image))
-
-        imglog.set(result=result)
-        _log_match_image_debug(imglog)
-
-        if grabbed_from_live:
-            self._display.draw(result, None)
-
-        if result.match:
-            debug("Match found: %s" % str(result))
-        else:
-            debug("No match found. Closest match: %s" % str(result))
-
-        return result
+            # pylint:disable=undefined-loop-variable
+            try:
+                for (matched, match_region, first_pass_matched,
+                     first_pass_certainty) in _find_matches(
+                        _crop(npframe, region), template.image,
+                        match_parameters, imglog):
+
+                    match_region = Region.from_extents(*match_region) \
+                                         .translate(region.x, region.y)
+                    result = MatchResult(
+                        get_frame_timestamp(frame), matched, match_region,
+                        first_pass_certainty, numpy.copy(npframe),
+                        (template.name or template.image))
+                    if grabbed_from_live:
+                        self._display.draw(result, None)
+                    yield result
+
+            except GeneratorExit:
+                imglog.set(first_pass_matched=first_pass_matched, result=result)
+                _log_match_image_debug(imglog)
+                raise
```

Comment:
Too subtle, I agree. I've force-pushed a change to use `finally` instead.

Originally I had used `except GeneratorExit` to avoid running the `imglog.set` in that block when `first_pass_matched` and `result` hadn't been defined if `_find_matches` raised an exception. Instead I've moved the `imglog.set` before the `yield` (the next commit did this, anyway, to log _all_ results from the loop).